src/Cms/ModelWithContent.php

```diff
@@ -169,59 +173,45 @@ public function content(string $languageCode = null)
 	}
 
 	/**
-	 * Returns the absolute path to the content file
-	 *
+	 * Returns the absolute path to the content file;
+	 * NOTE: only supports the published content file
+	 * (use `$model->storage()->contentFile()` for other versions)
 	 * @internal
-	 * @param string|null $languageCode
-	 * @param bool $force
-	 * @return string
+	 * @deprecated 4.0.0
+	 * @todo Remove in v5
+	 *
 	 * @throws \Kirby\Exception\InvalidArgumentException If the language for the given code does not exist
 	 */
 	public function contentFile(string $languageCode = null, bool $force = false): string
 	{
-		$extension = $this->contentFileExtension();
-		$directory = $this->contentFileDirectory();
-		$filename  = $this->contentFileName();
-
-		// overwrite the language code
-		if ($force === true) {
-			if (empty($languageCode) === false) {
-				return $directory . '/' . $filename . '.' . $languageCode . '.' . $extension;
-			}
-
-			return $directory . '/' . $filename . '.' . $extension;
-		}
-
-		// add and validate the language code in multi language mode
-		if ($this->kirby()->multilang() === true) {
-			if ($language = $this->kirby()->languageCode($languageCode)) {
-				return $directory . '/' . $filename . '.' . $language . '.' . $extension;
-			}
+		Helpers::deprecated('The internal $model->contentFile() method has been deprecated', 'model-content-file');
 
-			throw new InvalidArgumentException('Invalid language: ' . $languageCode);
-		}
+		$identifier = $this::CLASS_ALIAS === 'page' && $this->isDraft() === true ?
+			VersionIdentifier::changes() :
+			VersionIdentifier::published();
 
-		return $directory . '/' . $filename . '.' . $extension;
+		return $this->storage()->contentFile(
+			$identifier,
+			$this->storage()->languageCodeToObject($languageCode, $force)
+		);
 	}
 
 	/**
-	 * Returns an array with all content files
-	 *
-	 * @return array
+	 * Returns an array with all content files;
+	 * NOTE: only supports the published content file
+	 * (use `$model->storage()->contentFiles()` for other versions)
+	 * @deprecated 4.0.0
+	 * @todo Remove in v5
 	 */
 	public function contentFiles(): array
 	{
-		if ($this->kirby()->multilang() === true) {
-			$files = [];
-			foreach ($this->kirby()->languages()->codes() as $code) {
-				$files[] = $this->contentFile($code);
-			}
-			return $files;
-		}
+		Helpers::deprecated('The internal $model->contentFiles() method has been deprecated', 'model-content-file');
```

Comment:
There are no 1:1 replacements for most of them, they will be removed completely in v5. As all methods were previously internal, I don't expect anyone to have used them. The deprecation warnings were added out of "extra caution" to catch potential issues with sites/plugins easier.

The only methods that have replacements are `contentFile()` and `contentFiles()`. They are now `storage()->contentFile()` and `storage()->contentFiles()`. But those are also internal and may be removed in a future major version, so I'd like to avoid pointing site and plugin devs to them in the deprecation warning.

Reply:
Not anyone? Is it that safe - I would think someone that tried to implement some virtual page storage that's not our text files would have needed to use those, wouldn't they? For them, I think the deprecation message needs to point them into any direction they can go forward with. Just "tough luck" doesn't feel good.

Reply:
We'll see what our users actually need once they test alpha 4 with their sites and plugins. 🙂

Reply:
I don’t quite get what’s the issue of pointing them directly in a good direction, even avoid frustrations during the alpha.